internal/planner/planner_test.go

```diff
@@ -224,6 +224,42 @@ func TestPlannerHelloWorld(t *testing.T) {
 			note:    "relation unify",
 			queries: []string{`walk(input, [["foo", y], x])`},
 		},
+		{
+			note:    "else conflict-1",
+			queries: []string{`data.p.q`},
+			modules: []string{
+				`package p
+
+				q {
+					false
+				}
+				else = true {
+					true
+				}
+				q = false
+				`,
+			},
+		},
+		{
+			note:    "else conflict-2",
+			queries: []string{`data.p.q`},
+			modules: []string{
+				`package p
+
+				q {
+					false
+				}
+				else = false {
+					true
+				}
+				q {
+					false
+				}
+				else = true {
+					true
+				}`,
+			},
```

Comment:
Here, what we had before was indeed wrong:

### Before

```
| *ir.Funcs Funcs (1 funcs)
| | *ir.Func g0.data.p.q (2 params: [0 1], 3 blocks)
| | | *ir.Block Block (1 statements)
| | | | *ir.BlockStmt BlockStmt (2 blocks) &{0 0 0  }
| | | | | *ir.Block Block (5 statements)
| | | | | | *ir.MakeBooleanStmt &{Value:false Target:3 Location:{Index:0 Col:6 Row:4 file:module-0.rego text:false}}
| | | | | | *ir.MakeBooleanStmt &{Value:false Target:4 Location:{Index:0 Col:6 Row:4 file:module-0.rego text:false}}
| | | | | | *ir.NotEqualStmt &{A:3 B:4 Location:{Index:0 Col:6 Row:4 file:module-0.rego text:false}}
| | | | | | *ir.MakeBooleanStmt &{Value:true Target:5 Location:{Index:0 Col:5 Row:3 file:module-0.rego text:q}}
| | | | | | *ir.AssignVarOnceStmt &{Target:2 Source:5 Location:{Index:0 Col:5 Row:3 file:module-0.rego text:q}}
| | | | | *ir.Block Block (6 statements)
| | | | | | *ir.IsUndefinedStmt &{Source:2 Location:{Index:0 Col:5 Row:6 file:module-0.rego text:else = false}}
| | | | | | *ir.MakeBooleanStmt &{Value:true Target:3 Location:{Index:0 Col:6 Row:7 file:module-0.rego text:true}}
| | | | | | *ir.MakeBooleanStmt &{Value:false Target:4 Location:{Index:0 Col:6 Row:7 file:module-0.rego text:true}}
| | | | | | *ir.NotEqualStmt &{A:3 B:4 Location:{Index:0 Col:6 Row:7 file:module-0.rego text:true}}
| | | | | | *ir.MakeBooleanStmt &{Value:false Target:5 Location:{Index:0 Col:5 Row:6 file:module-0.rego text:else = false}}
| | | | | | *ir.AssignVarOnceStmt &{Target:2 Source:5 Location:{Index:0 Col:5 Row:6 file:module-0.rego text:else = false}}
| | | *ir.Block Block (1 statements)
| | | | *ir.BlockStmt BlockStmt (2 blocks) &{0 0 0  }
| | | | | *ir.Block Block (5 statements)
| | | | | | *ir.MakeBooleanStmt &{Value:false Target:3 Location:{Index:0 Col:6 Row:10 file:module-0.rego text:false}}
| | | | | | *ir.MakeBooleanStmt &{Value:false Target:4 Location:{Index:0 Col:6 Row:10 file:module-0.rego text:false}}
| | | | | | *ir.NotEqualStmt &{A:3 B:4 Location:{Index:0 Col:6 Row:10 file:module-0.rego text:false}}
| | | | | | *ir.MakeBooleanStmt &{Value:true Target:5 Location:{Index:0 Col:5 Row:9 file:module-0.rego text:q}}
| | | | | | *ir.AssignVarOnceStmt &{Target:2 Source:5 Location:{Index:0 Col:5 Row:9 file:module-0.rego text:q}}
| | | | | *ir.Block Block (6 statements)
| | | | | | *ir.IsUndefinedStmt &{Source:2 Location:{Index:0 Col:5 Row:12 file:module-0.rego text:else = true}}
| | | | | | *ir.MakeBooleanStmt &{Value:true Target:3 Location:{Index:0 Col:6 Row:13 file:module-0.rego text:true}}
| | | | | | *ir.MakeBooleanStmt &{Value:false Target:4 Location:{Index:0 Col:6 Row:13 file:module-0.rego text:true}}
| | | | | | *ir.NotEqualStmt &{A:3 B:4 Location:{Index:0 Col:6 Row:13 file:module-0.rego text:true}}
| | | | | | *ir.MakeBooleanStmt &{Value:true Target:5 Location:{Index:0 Col:5 Row:12 file:module-0.rego text:else = true}}
| | | | | | *ir.AssignVarOnceStmt &{Target:2 Source:5 Location:{Index:0 Col:5 Row:12 file:module-0.rego text:else = true}}
| | | *ir.Block Block (1 statements)
| | | | *ir.ReturnLocalStmt &{Source:2 Location:{Index:0 Col:5 Row:3 file:module-0.rego text:q}}
```

The second `else` block's `IsUndefinedStmt` checks the `fn.Return` local:
```
| | | | | | *ir.IsUndefinedStmt &{Source:2 Location:{Index:0 Col:5 Row:12 file:module-0.rego text:else = true}}
```
...which is **not undefined** after the first `else` block has been executed.


### After

```
| *ir.Funcs Funcs (1 funcs)
| | *ir.Func g0.data.p.q (2 params: [0 1], 3 blocks)
| | | *ir.Block Block (1 statements)
| | | | *ir.BlockStmt BlockStmt (3 blocks) &{0 0 0  }
| | | | | *ir.Block Block (5 statements)
| | | | | | *ir.MakeBooleanStmt &{Value:false Target:5 Location:{Index:0 Col:6 Row:4 file:module-0.rego text:false}}
| | | | | | *ir.MakeBooleanStmt &{Value:false Target:6 Location:{Index:0 Col:6 Row:4 file:module-0.rego text:false}}
| | | | | | *ir.NotEqualStmt &{A:5 B:6 Location:{Index:0 Col:6 Row:4 file:module-0.rego text:false}}
| | | | | | *ir.MakeBooleanStmt &{Value:true Target:7 Location:{Index:0 Col:5 Row:3 file:module-0.rego text:q}}
| | | | | | *ir.AssignVarStmt &{Source:7 Target:3 Location:{Index:0 Col:5 Row:3 file:module-0.rego text:q}}
| | | | | *ir.Block Block (6 statements)
| | | | | | *ir.IsUndefinedStmt &{Source:3 Location:{Index:0 Col:5 Row:6 file:module-0.rego text:else = false}}
| | | | | | *ir.MakeBooleanStmt &{Value:true Target:5 Location:{Index:0 Col:6 Row:7 file:module-0.rego text:true}}
| | | | | | *ir.MakeBooleanStmt &{Value:false Target:6 Location:{Index:0 Col:6 Row:7 file:module-0.rego text:true}}
| | | | | | *ir.NotEqualStmt &{A:5 B:6 Location:{Index:0 Col:6 Row:7 file:module-0.rego text:true}}
| | | | | | *ir.MakeBooleanStmt &{Value:false Target:7 Location:{Index:0 Col:5 Row:6 file:module-0.rego text:else = false}}
| | | | | | *ir.AssignVarStmt &{Source:7 Target:3 Location:{Index:0 Col:5 Row:6 file:module-0.rego text:else = false}}
| | | | | *ir.Block Block (2 statements)
| | | | | | *ir.IsDefinedStmt &{Source:3 Location:{Index:0 Col:5 Row:6 file:module-0.rego text:else = false}}
| | | | | | *ir.AssignVarOnceStmt &{Target:2 Source:3 Location:{Index:0 Col:5 Row:6 file:module-0.rego text:else = false}}
| | | *ir.Block Block (1 statements)
| | | | *ir.BlockStmt BlockStmt (3 blocks) &{0 0 0  }
| | | | | *ir.Block Block (5 statements)
| | | | | | *ir.MakeBooleanStmt &{Value:false Target:5 Location:{Index:0 Col:6 Row:10 file:module-0.rego text:false}}
| | | | | | *ir.MakeBooleanStmt &{Value:false Target:6 Location:{Index:0 Col:6 Row:10 file:module-0.rego text:false}}
| | | | | | *ir.NotEqualStmt &{A:5 B:6 Location:{Index:0 Col:6 Row:10 file:module-0.rego text:false}}
| | | | | | *ir.MakeBooleanStmt &{Value:true Target:7 Location:{Index:0 Col:5 Row:9 file:module-0.rego text:q}}
| | | | | | *ir.AssignVarStmt &{Source:7 Target:4 Location:{Index:0 Col:5 Row:9 file:module-0.rego text:q}}
| | | | | *ir.Block Block (6 statements)
| | | | | | *ir.IsUndefinedStmt &{Source:4 Location:{Index:0 Col:5 Row:12 file:module-0.rego text:else = true}}
| | | | | | *ir.MakeBooleanStmt &{Value:true Target:5 Location:{Index:0 Col:6 Row:13 file:module-0.rego text:true}}
| | | | | | *ir.MakeBooleanStmt &{Value:false Target:6 Location:{Index:0 Col:6 Row:13 file:module-0.rego text:true}}
| | | | | | *ir.NotEqualStmt &{A:5 B:6 Location:{Index:0 Col:6 Row:13 file:module-0.rego text:true}}
| | | | | | *ir.MakeBooleanStmt &{Value:true Target:7 Location:{Index:0 Col:5 Row:12 file:module-0.rego text:else = true}}
| | | | | | *ir.AssignVarStmt &{Source:7 Target:4 Location:{Index:0 Col:5 Row:12 file:module-0.rego text:else = true}}
| | | | | *ir.Block Block (2 statements)
| | | | | | *ir.IsDefinedStmt &{Source:4 Location:{Index:0 Col:5 Row:12 file:module-0.rego text:else = true}}
| | | | | | *ir.AssignVarOnceStmt &{Target:2 Source:4 Location:{Index:0 Col:5 Row:12 file:module-0.rego text:else = true}}
| | | *ir.Block Block (1 statements)
| | | | *ir.ReturnLocalStmt &{Source:2 Location:{Index:0 Col:5 Row:3 file:module-0.rego text:q}}
```

Now, we've got one local per set of "rule+elses", and each `else` only checks if the corresponding rule's local is undefined,

```
| | | | | | *ir.IsUndefinedStmt &{Source:3 Location:{Index:0 Col:5 Row:6 file:module-0.rego text:else = false}}
```
for the first else block (return local 3) and
```
| | | | | | *ir.IsUndefinedStmt &{Source:4 Location:{Index:0 Col:5 Row:12 file:module-0.rego text:else = true}}
```
for the second (return local 4).

Each rule+elses set gets one block for the rule, one per each, and one block for AssignVarOnceStmt, guarded by IsDefinedStmt.

Reply:
With the re-used local, we now have:

```
| *ir.Funcs Funcs (1 funcs)
| | *ir.Func g0.data.p.q (2 params: [0 1], 3 blocks)
| | | *ir.Block Block (1 statements)
| | | | *ir.BlockStmt BlockStmt (3 blocks) &{0 0 0  }
| | | | | *ir.Block Block (6 statements)
| | | | | | *ir.ResetLocalStmt &{Target:3 Location:{Index:0 Col:5 Row:3 file:module-0.rego text:q}}
| | | | | | *ir.MakeBooleanStmt &{Value:false Target:4 Location:{Index:0 Col:6 Row:4 file:module-0.rego text:false}}
| | | | | | *ir.MakeBooleanStmt &{Value:false Target:5 Location:{Index:0 Col:6 Row:4 file:module-0.rego text:false}}
| | | | | | *ir.NotEqualStmt &{A:4 B:5 Location:{Index:0 Col:6 Row:4 file:module-0.rego text:false}}
| | | | | | *ir.MakeBooleanStmt &{Value:true Target:6 Location:{Index:0 Col:5 Row:3 file:module-0.rego text:q}}
| | | | | | *ir.AssignVarStmt &{Source:6 Target:3 Location:{Index:0 Col:5 Row:3 file:module-0.rego text:q}}
| | | | | *ir.Block Block (6 statements)
| | | | | | *ir.IsUndefinedStmt &{Source:3 Location:{Index:0 Col:5 Row:6 file:module-0.rego text:else = false}}
| | | | | | *ir.MakeBooleanStmt &{Value:true Target:4 Location:{Index:0 Col:6 Row:7 file:module-0.rego text:true}}
| | | | | | *ir.MakeBooleanStmt &{Value:false Target:5 Location:{Index:0 Col:6 Row:7 file:module-0.rego text:true}}
| | | | | | *ir.NotEqualStmt &{A:4 B:5 Location:{Index:0 Col:6 Row:7 file:module-0.rego text:true}}
| | | | | | *ir.MakeBooleanStmt &{Value:false Target:6 Location:{Index:0 Col:5 Row:6 file:module-0.rego text:else = false}}
| | | | | | *ir.AssignVarStmt &{Source:6 Target:3 Location:{Index:0 Col:5 Row:6 file:module-0.rego text:else = false}}
| | | | | *ir.Block Block (2 statements)
| | | | | | *ir.IsDefinedStmt &{Source:3 Location:{Index:0 Col:5 Row:6 file:module-0.rego text:else = false}}
| | | | | | *ir.AssignVarOnceStmt &{Target:2 Source:3 Location:{Index:0 Col:5 Row:6 file:module-0.rego text:else = false}}
| | | *ir.Block Block (1 statements)
| | | | *ir.BlockStmt BlockStmt (3 blocks) &{0 0 0  }
| | | | | *ir.Block Block (6 statements)
| | | | | | *ir.ResetLocalStmt &{Target:3 Location:{Index:0 Col:5 Row:9 file:module-0.rego text:q}}
| | | | | | *ir.MakeBooleanStmt &{Value:false Target:4 Location:{Index:0 Col:6 Row:10 file:module-0.rego text:false}}
| | | | | | *ir.MakeBooleanStmt &{Value:false Target:5 Location:{Index:0 Col:6 Row:10 file:module-0.rego text:false}}
| | | | | | *ir.NotEqualStmt &{A:4 B:5 Location:{Index:0 Col:6 Row:10 file:module-0.rego text:false}}
| | | | | | *ir.MakeBooleanStmt &{Value:true Target:6 Location:{Index:0 Col:5 Row:9 file:module-0.rego text:q}}
| | | | | | *ir.AssignVarStmt &{Source:6 Target:3 Location:{Index:0 Col:5 Row:9 file:module-0.rego text:q}}
| | | | | *ir.Block Block (6 statements)
| | | | | | *ir.IsUndefinedStmt &{Source:3 Location:{Index:0 Col:5 Row:12 file:module-0.rego text:else = true}}
| | | | | | *ir.MakeBooleanStmt &{Value:true Target:4 Location:{Index:0 Col:6 Row:13 file:module-0.rego text:true}}
| | | | | | *ir.MakeBooleanStmt &{Value:false Target:5 Location:{Index:0 Col:6 Row:13 file:module-0.rego text:true}}
| | | | | | *ir.NotEqualStmt &{A:4 B:5 Location:{Index:0 Col:6 Row:13 file:module-0.rego text:true}}
| | | | | | *ir.MakeBooleanStmt &{Value:true Target:6 Location:{Index:0 Col:5 Row:12 file:module-0.rego text:else = true}}
| | | | | | *ir.AssignVarStmt &{Source:6 Target:3 Location:{Index:0 Col:5 Row:12 file:module-0.rego text:else = true}}
| | | | | *ir.Block Block (2 statements)
| | | | | | *ir.IsDefinedStmt &{Source:3 Location:{Index:0 Col:5 Row:12 file:module-0.rego text:else = true}}
| | | | | | *ir.AssignVarOnceStmt &{Target:2 Source:3 Location:{Index:0 Col:5 Row:12 file:module-0.rego text:else = true}}
| | | *ir.Block Block (1 statements)
| | | | *ir.ReturnLocalStmt &{Source:2 Location:{Index:0 Col:5 Row:3 file:module-0.rego text:q}}
```

Note that it's resetting the local `3` in each block.